File: AdventOfCode.2024/Day_02/Day02_Part2.cs
namespace AdventOfCode2024;

internal sealed class Day02_Part2 : PuzzleSolution
{
    public static string FileName { get; } = "Day_02/Input.txt";
    public static string TestFileName { get; } = "Day_02/Example.txt";
    public static string TestOutputExpected { get; } = "4";

    public static string Solve(StreamReader reader)
    {
        var safeReports = 0;
        while (reader.ReadLine() is { } line)
        {
            var numbers = line.Split(' ').Select(int.Parse).ToArray();
            var reportIsSafe = false;

            for (var i = 0; i < numbers.Length; i++)
            {
                var numbersWithoutIndex = numbers.Where((_, index) => index != i);
                reportIsSafe = ReportIsSafe(numbersWithoutIndex.ToArray());
                if (reportIsSafe)
                {
                    safeReports++;
                    break;
                }
            }
        }

        return safeReports.ToString();
    }

    private static bool ReportIsSafe(int[] numbers)
    {
        var direction = Change.None;
        for (var i = 1; i < numbers.Length; i++)
        {
            var diff = numbers[i] - numbers[i - 1];
            var isWithinRange = diff != 0 && Math.Abs(diff) <= 3;
            if (isWithinRange is false)
            {
                return false;
            }
            if (direction == Change.None)
            {
                direction = diff > 0 ? Change.Increasing : Change.Decreasing;
            }
            if (direction == Change.Increasing && diff < 0)
            {
                return false;
            }
            if (direction == Change.Decreasing && diff > 0)
            {
                return false;
            }
        }
        return true;
    }

    private enum Change
    {
        None = 0,
        Increasing = 1,
        Decreasing = -1
    }
}

File: javascript/2024/solutions/day02part2.js
export const inputFile = '02.input';
export const testFile = '02.sample';
export const testOutputExpected = "4";

export function solve(input) {
    let safeReports = 0;

    input.split('\n').forEach(line => {
        let report = line.split(' ').map(x => parseInt(x));
        for (let i = 0; i < report.length; i++) {
            let reportCopy = report.filter((_, index) => index !== i);
            let isSafe = reportIsSafe(reportCopy) ? 1 : 0;
            if (isSafe) {
                safeReports++;
                break;
            }
        }
    });
    
    return safeReports.toString();
}

function reportIsSafe(report) {
    let direction = 0;
    for (let i = 1; i < report.length; i++) {
        let diff = report[i] - report[i - 1];
        let isWithinRange = diff !== 0 && Math.abs(diff) <= 3;
        if (!isWithinRange) {
            return false;
        }
        if (direction === 0) {
            direction = diff > 0 ? 1 : -1;
            continue;
        }
        if (direction === 1 && diff < 0) {
            return false;
        }
        if (direction === -1 && diff > 0) {
            return false;
        }
    }
    return true;
}